grouper/models/public_key.py

```diff
@@ -18,3 +75,35 @@ class PublicKey(Model):
     public_key = Column(Text, nullable=False)
     fingerprint = Column(String(length=64), nullable=False, unique=True)
     created_on = Column(DateTime, default=datetime.utcnow, nullable=False)
+
+    def tags(self):
+        # type: () -> List[PublicKeyTag]
+        """Returns the list of tags that are assigned to this public key
+
+        Returns:
+            a list that contains all of the PublicKeyTags that are assigned to this public key
+        """
+        mappings = self.session.query(PublicKeyTagMap).filter_by(key_id=self.id).all()
+        return [mapping.tag for mapping in mappings]
+
+    def my_permissions(self):
+        # type: () -> List[Permission]
+        """Returns the permissions that this public key has. Namely, this the set of permissions
+        that the public key's owner has, intersected with the permissions allowed by this key's
+        tags
+
+        Returns:
+            a list of all permissions this public key has
+        """
+        my_perms = self.user.my_permissions()
+        for tag in self.tags():
+            my_perms = _permission_intersection(my_perms, tag.my_permissions())
+
+        return list(my_perms)
+
+    def pretty_my_permissions(self):
```

Comment:
I talked with @lfaraone about this, and we weren't sure on the best way to approach this. The biggest problem is the lack of a clean way (that I'm familiar with at least) to do the equivalent functionality in Jinja. Additionally, because the use case for this (https://github.com/dropbox/grouper/pull/208/files#diff-64a1bd8aadee752b60baa65ad62066a8R31) requires that it be run for every key that a user has, if we wanted to move this to another part of the code and run it in say the handler to pass into the template, we'd have to have some way of coupling the output with the PermissionKey objects themselves. Do you have a particular solution in mind?